docs/automatic-publishing-to-s3.md

```diff
@@ -0,0 +1,34 @@
+# Automatic publishing of a static site to S3
+
+[Wagtail Bakery](https://github.com/wagtail/wagtail-bakery) can build ("bake") a static version of the site and sync it to Amazon S3, where it can be served as a website.
+
+_Note that "Publish" and "Unpublish" here refer to the Wagtail sense of making a page live / not live. Emitting a version of the site to S3 will be referred to as "build-and-sync"._
+
+In production, this "build-and-sync" process will run at a few points:
+
+1. It will be run whenever a Page is Published or Unpublished from the Wagtail admin. (Note: this means if two Pages are published, the "build-and-sync" process is run twice...)
+
+2. Hourly, on a schedule, we'll check for any Wagtail Pages due to be automaticaly Published or Unpublished, and if any are found, that will be done. This action will then trigger the "build-and-sync" step mentioned in 1 for each page being automatically Published or Unpublished.
+
+3. Hourly, on a schedule, the whole site needs to republished, so that listings of upcoming Events, for instance, don't become too stale (especially if it's a weekend, for instance).
+
+**Case 1** will be performed asynchronously, via a task queue. Publishing/Unpublishing a Page immediately puts a task in the queue. A worker process picks it up and runs the "build-and-sync" process. This process will sync new HTML for **every Page within the site** as well as any new/changed assets (CSS, JS, images and user media) to the S3 bucket.
+
+**Case 2** will be triggered by a scheduler/cron job running Wagtail's [`publish_scheduled_pages`](https://docs.wagtail.io/en/v2.0/reference/management_commands.html#publish-scheduled-pages) management command. This implicitly then triggers `Case 1`.
+
+**Case 3** will be triggered by a different scheduler/cron job. How we implement this is unconfirmed at the moment. Two options are:
+
+1. Schedule a job that runs in the main app container and executes the two relevant wagtail-bakery commands:
+
+   ```
+   /usr/local/bin/python /app/manage.py build &&
+   /usr/local/bin/python /app/manage.py publish
+   ```
+
+This will be synchronous, but should not interfere with the app's request-response cycle. We could probably also do this in the `worker` container.
+
+2. We add a new management command that emits the post-Publish Django Signal that starts **Case 1**, so we end up with a new queued task, which the worker picks up.
+
+## A note about queue load
+
+There is a risk that the various ways of triggering an async build, plus the fact that each Publish/Unpublish action for each page will trigger a new static build, means there may be a lot of "build-and-sync" tasks in the queue, and all but the most recent one will be redundant. (We might want to look at a way to purge outmoded "build-and-sync" tasks and just use the newest one.)
```

Comment:
> We might want to look at a way to purge outmoded "build-and-sync" tasks and just use the newest one.

I think this is the way to go -- what does everyone else think?

Reply:
I think so too, that sounds like a nice feature to have!